tests/z3-tests.adb

```diff
@@ -813,6 +813,46 @@ package body Z3.Tests is
 
    ---------------------------------------------------------------------------
 
+   procedure Test_Reference_Counting (T : in out Test_Cases.Test_Case'Class)
+   is
+      pragma Unreferenced (T);
+      Context  : Z3.Context;
+      Solver   : constant Z3.Solver   := Create (Context);
+      Optimize : constant Z3.Optimize := Create (Context);
+
+      procedure Copy_Ctx (Local_Ctx : Z3.Context)
+      is
+         Ctx2          : Z3.Context;
+         Solver        : Z3.Solver := Create (Context);
+         Unused_Result : Result;
+      begin
+         Ctx2 := Local_Ctx;
+         pragma Inspection_Point (Ctx2);
+         Solver.Assert (Bool ("A", Context));
+         Unused_Result := Solver.Check;
+      end Copy_Ctx;
+      procedure Copy_Solver (Local_Solver : Z3.Solver)
+      is
+         Solver2 : Z3.Solver := Create (Context);
+      begin
+         Solver2 := Local_Solver;
+         pragma Inspection_Point (Solver2);
+      end Copy_Solver;
+      procedure Copy_Optimize (Local_Optimize : Z3.Optimize)
+      is
+         Optimize2 : Z3.Optimize := Create (Context);
+      begin
+         Optimize2 := Local_Optimize;
+         pragma Inspection_Point (Optimize2);
+      end Copy_Optimize;
+   begin
+      Copy_Ctx (Context);
+      Copy_Solver (Solver);
+      Copy_Optimize (Optimize);
```

Comment:
What is being tested here (I don't see any assertion)?

Reply:
This tests the reference counting by creating copies in procedures and letting them go out of scope. The test is checked by the address sanitizer. We don't need AUnit for this test but it doesn't make sense to create a separate test binary just for that.

Reply:
I don't see a problem with having it here, but we should add a comment to remind our future selves of that.

Reply:
Fixed by https://github.com/Componolit/AZ3/pull/35/commits/476de4c4fd0e949efae48fc1b63a428fbef58423.